local-volume/provisioner/pkg/deleter/blockcleaner.go

```diff
@@ -0,0 +1,247 @@
+		// Since we are trying to fill the block device, a healthy end to the execution would be the error
+		// message "No space left on device".
+		args := []string{"if=" + ZeroDev, "of=" + mountPath, "bs=" + DefaultBlockSize, "status=progress"}
+		return iters, DDFullErrExpected, DDCommand, args, nil
+
+	case common.BlockCleanupRandom:
+		// shred -n 3 /dev/sda # we defer to shred to run the iterations
+		return 1, NoErrExpected, ShredCommand, []string{"-vzf", "-n", strconv.Itoa(iters), mountPath}, nil
+
+	case common.BlockCleanupCustom:
+		// sh -C "custom_cmd /dev/sda"
+		return iters, NoErrExpected, ShellCommand, []string{"-c", config.BlockConf.CustomExec + " " + mountPath}, nil
+	}
+
+	return 0, NoErrExpected, "", []string{}, fmt.Errorf("Unknown BlockCleanerImpl type %q specified in config",
+		config.BlockConf.Cleaner)
+}
+
+func (v *BlockCleanerImpl) asyncDeleteBlockPV(pv *v1.PersistentVolume, iterations int, expectErrorOut string,
+	exe string, exeArgs ...string) error {
+	err := v.execAllIterations(pv.Name, iterations, expectErrorOut, exe, exeArgs...)
+	if err != nil {
+		glog.Error(err)
+		return err
+	}
+
+	// Remove API object
+	if err := v.APIUtil.DeletePV(pv.Name); err != nil {
+		glog.Error(err)
+		deletingLocalPVErr := fmt.Errorf("Error deleting block PV %q: %v", pv.Name, err.Error())
+		v.RuntimeConfig.Recorder.Eventf(pv, v1.EventTypeWarning, common.EventVolumeFailedDelete,
+			deletingLocalPVErr.Error())
+		return err
+	}
+
+	return nil
+}
+
+func (v *BlockCleanerImpl) execAllIterations(pvName string, iterations int, expectErrOut string,
+	exe string, exeArgs ...string) error {
+
+	for i := 0; i < iterations; i++ {
+		if i > 0 {
+			glog.Infof("PV %s : Cleaner %s %s", pvName, exe, exeArgs)
+		}
+		glog.Infof("%d) Cleaning PV %q: Executing %s", i, pvName, exeArgs)
+		err := v.execScript(pvName, expectErrOut, exe, exeArgs...)
+		if err != nil {
+			glog.Errorf("Cleaning PV %q: Executing %s %s failed with error %v", pvName, exe, exeArgs, err)
+			v.markDone(pvName)
+			return err
+		}
+		glog.Infof("%d) Cleaning PV %q: Completed execution %s %s", i, pvName, exe, exeArgs)
+	}
+	v.markDone(pvName)
+	return nil
+}
+
+func (v *BlockCleanerImpl) execScript(pvName string, expectErrOut string, exe string, exeArgs ...string) error {
+	cmd := exec.Command(exe, exeArgs...)
+	var wg sync.WaitGroup
+	wg.Add(2)
+	var errBuf *bytes.Buffer
+
+	if expectErrOut != "" {
+		errBuf = bytes.NewBufferString("")
+	}
+
+	outReader, err := cmd.StdoutPipe()
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		defer wg.Done()
+		outScanner := bufio.NewScanner(outReader)
+		for outScanner.Scan() {
+			glog.Infof("Cleaner %q stdout: %q", pvName, outScanner.Text())
+		}
+	}()
+
+	errReader, err := cmd.StderrPipe()
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		defer wg.Done()
+		errScanner := bufio.NewScanner(errReader)
+		for errScanner.Scan() {
+			glog.Infof("Cleaner %q stderr: %q", pvName, errScanner.Text())
```

Comment:
Cleaner pvname: %q

Reply:
ok

Reply:
OK, I made the log messages more consistent, but a bit different from above.